GEODE-5591: handle IO exception and try with a random port, then retr…

gesterzhou opened the pull request:
…y with next port

@jhuynh1 @nabarunnag @boglesby 
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

gesterzhou commented:
To improve the 2-minute-timeout is another ticket. 

GEODE-5591 is only to resolve following 2 issues:
1) exception handling
2) retry next port. The old old has the risk to retry forever.

jhuynh1 commented on geode-wan/src/main/java/org/apache/geode/internal/cache/wan/GatewayReceiverImpl.java:
```diff
@@ -138,6 +136,29 @@ public CacheServer getServer() {
   }
 
+  private boolean tryToStart() {
+    if (!AvailablePort.isPortAvailable(port, AvailablePort.SOCKET,
```
This will possibly help but there still is the initial problem where if multiple receivers get the same start port, they all can check available port and all return true before one of them actually permanently binds.  I guess the issue is that we still have to wait 2 minutes if two ports are the same

gesterzhou replied:
yes, that's possible. As I said, it will be improved by reducing 2 miniute to a few seconds, or other solution in another  ticket.

jhuynh1 commented on geode-wan/src/main/java/org/apache/geode/internal/cache/wan/GatewayReceiverImpl.java:
```diff
@@ -146,53 +167,37 @@ public void start() throws IOException {
-        }
+    while (!started) {
+      this.port = port;
+      started = tryToStart();
```
maybe pass in the port to tryToStart(int port).  Then you can assign this.port after things are bound.  Not a big deal but might make it easier to read

gesterzhou replied:
I purposely do it this way, otherwise the exception message will not show the real port it finally tried.

jhuynh1 commented on geode-wan/src/main/java/org/apache/geode/internal/cache/wan/GatewayReceiverImpl.java:
```diff
@@ -146,53 +167,37 @@ public void start() throws IOException {
-        } else {
-          throw se;
-        }
+    while (!started) {
```
this might as well be a while(true).  Anytime started is true, we already break out.  Maybe we could condense the looping logic with the start/end and loop port logic, that might condense the code a bit (not sure if it's any better other than possibly less lines)

mcmellawatt commented:
Hi Gester,

I had an idea for another approach which keeps the simplicity of @smgoller's original fix but addresses the hanging during the 2 minute retry in AcceptorImpl.  We can simply just bypass the 2 minute retry logic if the CacheServerImpl is a GatewayReceiverImpl.  There is already a flag for this passed into the AcceptorImpl constructor.  Check out my branch:

https://github.com/mcmellawatt/geode/tree/feature/GEODE-5591

It is possible that if say 20 gateway receivers are being created at the same time that we will get 19 IOExceptions (they are all fighting for the same port), but they will fail fast and move onto the next port in the range.  I don't think this would impact the user experience at all, just something to keep in mind.

gesterzhou commented:
Let 20 members to compete for the same port definitely is a wrong idea. So
I fixed this part by put back the random port (as first try)

However, to skip timeout part is great. I apply this changes and added
test. I have worked out a new Pull-Request.

On Wed, Sep 5, 2018 at 5:18 PM, Ryan McMahon <notifications@github.com>
wrote:

> Hi Gester,
>
> I had an idea for another approach which keeps the simplicity of @smgoller
> <https://github.com/smgoller>'s original fix but addresses the hanging
> during the 2 minute retry in AcceptorImpl. We can simply just bypass the 2
> minute retry logic if the CacheServerImpl is a GatewayReceiverImpl. There
> is already a flag for this passed into the AcceptorImpl constructor. Check
> out my branch:
>
> https://github.com/mcmellawatt/geode/tree/feature/GEODE-5591
>
> It is possible that if say 20 gateway receivers are being created at the
> same time that we will get 19 IOExceptions (they are all fighting for the
> same port), but they will fail fast and move onto the next port in the
> range. I don't think this would impact the user experience at all, just
> something to keep in mind.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/geode/pull/2425#issuecomment-418922324>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKN7Xtp3TUTyD53O4Xd-F6E-0Z9oxzPDks5uYGnugaJpZM4WbzYa>
> .
>

mcmellawatt commented:
Hey Gester,

At second glance, I see this comment regarding the retry logic in AcceptorImpl:

```
// fix for bug 36617. If BindException is thrown, retry after
// sleeping. The server may have been stopped and then	         
// immediately restarted, which sometimes results in a bind exception	       
```

This makes me a bit nervous about bypassing that logic in a GatewayReceiverImpl context, since it is just a subclass of CacheServerImpl and may be subject to the same issue of encountering bind exceptions on server restart.

I'd vote we remove the changes I suggested in AcceptorImpl after looking at this more closely.  Your other changes should reduce the chance of collision dramatically so I think it's unlikely a customer would encounter the 2 minute retry.

mcmellawatt commented on geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java:
```diff
@@ -458,7 +458,8 @@ public AcceptorImpl(int port, String bindHostName, boolean notifyBySubscription,
             break;
           } catch (SocketException b) {
-            if (!treatAsBindException(b) || System.currentTimeMillis() > tilt) {
+            if (isGatewayReceiver || !treatAsBindException(b)
```
See my other comment - I think that this retry logic is protecting us against a specific case when the server restarts and encounters a bind exception, so it is probably wise to leave this file how it was.

gesterzhou replied:
OK. It's fair. I think it will not encounter that easily. I will remove the fix for this, including my new test.

mcmellawatt replied:
Thanks Gester, sorry for leading you astray ;).

gesterzhou commented:
Yes, I worked out a new diff to revoke this part of fix.

On Thu, Sep 6, 2018 at 9:52 AM, Ryan McMahon <notifications@github.com>
wrote:

> *@mcmellawatt* requested changes on this pull request.
>
> After a closer look, I think we should not bypass the retry logic for
> gateway receivers. See my comment above.
> ------------------------------
>
> In geode-core/src/main/java/org/apache/geode/internal/cache/
> tier/sockets/AcceptorImpl.java
> <https://github.com/apache/geode/pull/2425#discussion_r215698529>:
>
> > @@ -458,7 +458,8 @@ public AcceptorImpl(int port, String bindHostName, boolean notifyBySubscription,
>              this.serverSock.bind(new InetSocketAddress(getBindAddress(), port), backLog);
>              break;
>            } catch (SocketException b) {
> -            if (!treatAsBindException(b) || System.currentTimeMillis() > tilt) {
> +            if (isGatewayReceiver || !treatAsBindException(b)
>
> See my other comment - I think that this retry logic is protecting us
> against a specific case when the server restarts and encounters a bind
> exception, so it is probably wise to leave this file how it was.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/geode/pull/2425#pullrequestreview-153022519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKN7XhUnBeAthfDGnZU-svzueTzfZiAdks5uYVLcgaJpZM4WbzYa>
> .
>
